devices/src/virtio/block.rs

```diff
@@ -158,16 +146,54 @@ struct Request {
     status_addr: GuestAddress,
 }
 
+/// The request header represents the mandatory fields of each block device request.
+///
+/// A request header contains the following fields:
+///   * request_type: an u32 value mapping to a read, write or flush operation.
+///   * reserved: 32 bits are reserved for future extensions of the Virtio Spec.
+///   * sector: an u64 value representing the offset where a read/write is to occur.
+///
+/// The header simplifies reading the request from memory as all request follow
+/// the same memory layout.
+#[derive(Copy, Clone)]
+#[repr(C)]
+struct RequestHeader {
+    request_type: u32,
+    _reserved: u32,
+    sector: u64,
+}
+
+// Safe because RequestHeader only contains plain data.
+unsafe impl DataInit for RequestHeader {}
+
+impl RequestHeader {
+    /// Reads the request header from GuestMemory starting at `addr`.
+    ///
+    /// Virtio 1.0 specifies that the data is transmitted by the driver in little-endian
+    /// format. Firecracker currently runs only on little endian platforms so we don't
+    /// need to do an explicit little endian read as all reads are little endian by default.
+    /// When running on a big endian platform, this code should not compile, and support
+    /// for explicit little endian reads is required.
+    #[cfg(target_endian = "little")]
+    fn read_from(memory: &GuestMemory, addr: GuestAddress) -> result::Result<Self, Error> {
+        let request_header: RequestHeader = memory
+            .read_obj_from_addr(addr)
+            .map_err(Error::GuestMemory)?;
+        Ok(request_header)
+    }
+}
+
 impl Request {
     fn parse(avail_desc: &DescriptorChain, mem: &GuestMemory) -> result::Result<Request, Error> {
         // The head contains the request type which MUST be readable.
         if avail_desc.is_write_only() {
             return Err(Error::UnexpectedWriteOnlyDescriptor);
         }
 
+        let request_header = RequestHeader::read_from(mem, avail_desc.addr)?;
         let mut req = Request {
-            request_type: request_type(&mem, avail_desc.addr)?,
-            sector: sector(&mem, avail_desc.addr)?,
+            request_type: RequestType::from(request_header.request_type),
```

Comment:
nit: Would it make sense renaming the `request_header.request_type` field to `request_header.virtio_request_type`?

I do not have a strong preference here, but I think it is a bit misleading, when reading the code, to transform a variable named `request_type`, which behind it is represented by a `VIRTIO_BLK_T_X` type, to a RequestType enum. The confusion comes from the similar naming, which would suggest that there is no difference, and I think it will be perfectly intelligible if the field name would encapsulate also what request type we have in the RequestHeader struct.

Reply:
This is definitely a valid point. Unfortunately we cannot simply change the name of the request_header.request_type field to virtio_request_type field because it breaks the assumption that the same fields specified in the Virtio specification are found in the Firecracker Virtio specification as well.

It makes a lot of sense to remove `request_type` and `sector` from Request and have the `request_header` field there instead, but this is a rather big change that is impacted by future refactoring that is required for separating handling the queues, the transport and the backend of devices.